transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java

```diff
@@ -32,8 +32,13 @@ final void receivedRdHup() {
         receivedRdHup = true;
     }
 
+    final boolean isReceivedRdHup() {
+        return receivedRdHup;
+    }
+
     boolean maybeMoreDataToRead() {
-        return isEdgeTriggered && lastBytesRead() > 0;
+        // If EPOLLRDHUP has been received we must read until we get a read error.
+        return isEdgeTriggered && (lastBytesRead() > 0 || receivedRdHup);
```

Comment:
`receivedRdHup` is typically only true as the socket is being closed so the `lastBytesRead() > 0` condition is more likely to be the determining factor in the conditional.